### PR TITLE
Adding Logging for log in and log out

### PR DESCRIPTION
**Pull request by ccaprio97**
## What’s changing?

Adding Logging for log in and log out

## Why?

[_1491_](https://e-jira.fns.usda.net/jira-prod/browse/NAC-1491)

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [ ] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: query-tool/src/Piipan.QueryTool/Pages/SignedOut.cshtml.cs [hunk 1]
```diff
@@ -1,4 +1,7 @@
 using System;
+using Microsoft.AspNetCore.Http;
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/Pages/SignedOut.cshtml.cs:2]: Nitpick: Were any of the changes in this file actually necessary?

File: query-tool/tests/Piipan.QueryTool.Tests/SignedOutTest.cs [hunk 2]
```diff
@@ -37,4 +38,4 @@ public async Task TestMessageOnGet()
             Assert.Contains("<h1>You are signed out.</h1>", output);
         }
     }
-}
\ No newline at end of file
+}
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Tests/SignedOutTest.cs:41]: Nitpick: Are any of the changes in this file actually needed?

Comment (ccaprio97) [reply query-tool/tests/Piipan.QueryTool.Tests/SignedOutTest.cs:41]: So I originally changed this file then with the new approach didn't need them and tried to fix the tests failing while in the car by editing the file on git on my phone

File: shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs [hunk 3]
```diff
@@ -4,13 +4,23 @@
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
 using Moq;
+using Piipan.Shared.Claims;
```
Comment (bgfoshay) [inline shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:9]: Question: Do any of the tests in this file verify the sign-in & sign-out logger messages? If not, we should add/modify tests to verify them appropriately.

Comment (ccaprio97) [reply shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:10]: I wrote a test for log out that adds the headers and does go into the new code block. I'm not sure if I'm testing the logging messages correctly. As for log in its really just the log statement on the existing code so it would be a part of the new session test that already existed

Comment (bgfoshay) [reply shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:10]: So we have a test InvokeAsync_NewSession, which is effectively a Sign-in test. But there's no logging verification for the new line added to the middleware -> logger.LogInformation($"[User logged in: {loggedInUsersName}"); We should verify that's getting logged. 

Maybe I'm missing it, but I don't see the sign-out test? I was thinking you'd add a test like InvokeAsync_SignOut where you set the context.Request.Headers["Referer"] to contain "logout", and then once executed, you verify that logger.LogInformation($"[User logged out: {loggedInUsersName}"); is called.

File: query-tool/tests/Piipan.QueryTool.Tests/SignedOutTest.cs (hunk 2 above)
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Tests/SignedOutTest.cs]: Nitpick (non-blocking): you removed the newline at the end of the file

File: shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs [hunk 4]
```diff
@@ -119,5 +136,72 @@ public async void InvokeAsync_ReturningSession()
                 It.IsAny<Exception>(),
                 (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()), Times.Never());
         }
+
+        [Fact]
+        public async void InvokeAsync_LogOutHeader()
```
Comment (ccaprio97) [inline shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:141]: This is the log out test that has the referer header

File: shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs (hunk 3 above)
Comment (ccaprio97) [reply shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:10]: ya thats the InvokeAsync_LogOutHeader test i just wasnt sure how to check the logger messages and when looking into it had lots of problems. Maybe we could pair program and figure it out really quick but running it manually i was able to see that it worked properly

File: shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs (hunk 4 above)
Comment (bgfoshay) [reply shared/tests/Piipan.Shared.Tests/Logging/AuthenticationLoggingMiddlewareTests.cs:141]: 👍 Yep, this is the test I was looking for. I'll give you a call tomorrow and we can pair on the verification piece of it.

**Review by bgfoshay — APPROVED** [review]
👍 Looks good. Thanks for adding the extra logic and the additional tests.
